File: demos/iperf3/kind/simple-test.go
// ###############################################################
// Name: Simple iperf3  test
// Desc: create 2 kind clusters :
// 1) MBG and iperf3 client
// 2) MBG and iperf3 server
// ##############################################################
package main

import (
	"os"
	"strconv"
	"time"

	"github.ibm.com/mbg-agent/pkg/util"

	log "github.com/sirupsen/logrus"
	"github.ibm.com/mbg-agent/pkg/api"
	"github.ibm.com/mbg-agent/pkg/client"
	mbgAux "github.ibm.com/mbg-agent/demos/utils"
	kindAux "github.ibm.com/mbg-agent/demos/utils/kind"
)

const (
	mbgCaCrt = "./mtls/ca.crt"
	//MBG1 parameters
	mbg1DataPort   = "30001"
	mbg1cPort      = uint16(30443)
	mbg1cPortLocal = "443"
	mbg1crt        = "./mtls/mbg1.crt"
	mbg1key        = "./mtls/mbg1.key"
	mbg1Name       = "mbg1"
	gwctl1Name     = "gwctl1"
	mbg1cni        = "default"
	srcSvc         = "iperf3-client"

	//MBG2 parameters
	mbg2DataPort   = "30001"
	mbg2cPort      = uint16(30443)
	mbg2cPortLocal = "443"
	mbg2crt        = "./mtls/mbg2.crt"
	mbg2key        = "./mtls/mbg2.key"
	mbg2Name       = "mbg2"
	gwctl2Name     = "gwctl2"
	mbg2cni        = "default"
	destSvc        = "iperf3-server"
	destPort       = uint16(5000)
	kindDestPort   = "30001"
)

var (
	mtlsFolder string = mbgAux.ProjDir + "/demos/utils/"
	folCl      string = mbgAux.ProjDir + "/demos/iperf3/manifests/iperf3-client"
	folSv      string = mbgAux.ProjDir + "/demos/iperf3/manifests/iperf3-server"
)

func main() {
	// call a Python function
	dataplane := "mtls"
	nologfile := false
	mbgAux.SetLog()
	log.Println("Working directory", mbgAux.ProjDir)
	//exec.chdir(proj_dir)
	//clean
	log.Print("Clean old kinds")
	mbgAux.RunCmd("make clean-kind")

	// build docker environment
	mbgAux.PrintHeader("Build docker image")
	mbgAux.RunCmd("make docker-build")
	kindAux.CreateKindMbg(mbg1Name, dataplane, nologfile)
	kindAux.CreateKindMbg(mbg2Name, dataplane, nologfile)

	// //get parameters
	mbg1Ip, _ := kindAux.GetKindIp(mbg1Name)
	mbg2Ip, _ := kindAux.GetKindIp(mbg2Name)

	//set gwctl
	parsedCertData, err := util.ParseTLSFiles(mtlsFolder+mbgCaCrt, mtlsFolder+mbg1crt, mtlsFolder+mbg1key)
	if err != nil {
		log.Error(err)
		os.Exit(1)
	}
	gwctl1 := client.New(mbg1Ip, mbg1cPort, parsedCertData.ClientConfig(mbg1Name))

	parsedCertData, err = util.ParseTLSFiles(mtlsFolder+mbgCaCrt, mtlsFolder+mbg2crt, mtlsFolder+mbg2key)
	if err != nil {
		log.Error(err)
		os.Exit(1)
	}
	gwctl2 := client.New(mbg1Ip, mbg2cPort, parsedCertData.ClientConfig(mbg2Name))

	//Add Peer
	mbgAux.PrintHeader("Add peers and send hello")
	gwctl1.Peers.Create(&api.Peer{Name: mbg2Name, Spec: api.PeerSpec{Gateways: []api.Endpoint{{Host: mbg2Ip, Port: mbg2cPort}}}})

	//Set iperf3 client
	mbgAux.PrintHeader("Add iperf3 client")
	kindAux.CreateServiceInKind(mbg1Name, srcSvc, "mlabbe/iperf3", folCl+"/"+srcSvc+".yaml")
	srcSvcPod, _ := mbgAux.GetPodNameIp(srcSvc)
	//gwctl1.AddService(srcSvc, "", "", "iperf3 client") //Allow to use all by default

	//Set iperf3 server
	mbgAux.PrintHeader("Add iperf3 server")
	kindAux.CreateServiceInKind(mbg2Name, destSvc, "mlabbe/iperf3", folSv+"/iperf3.yaml")
	destSvcPod, destSvcIP := mbgAux.GetPodNameIp(destSvc)

	gwctl2.Exports.Create(&api.Export{Name: destSvc, Spec: api.ExportSpec{Service: api.Endpoint{Host: destSvcIP, Port: destPort}}})
	log.Println(srcSvcPod, destSvcPod)

	//Expose service
	mbgAux.PrintHeader("Start expose")
	kindAux.UseKindCluster(mbg2Name)
	gwctl2.Imports.Create(&api.Import{Name: destSvc, Spec: api.ImportSpec{Service: api.Endpoint{Host: destSvc, Port: destPort}}})

	//bindK8sSvc()
	mbgAux.PrintHeader("Bind a service")
	kindAux.UseKindCluster(mbg1Name)
	gwctl1.Bindings.Create(&api.Binding{Spec: api.BindingSpec{Import: destSvc, Peer: mbg2Name}})
	time.Sleep(5 * time.Second)

	//iperf3test
	mbgAux.RunCmdNoPipe("kubectl exec -i " + srcSvcPod + " -- iperf3 -c " + destSvc + " -p " + strconv.Itoa(int(destPort)))

}

// ############################### MAIN ##########################
// if __name__ == "__main__":
//     parser = argparse.ArgumentParser(description='Description of your program')
//     parser.add_argument('-d','--dataplane', help='choose which dataplane to use mtls/tcp', required=False, default="mtls")
//     parser.add_argument('-c','--cni', help='Which cni to use default(kindnet)/flannel/calico/diff (different cni for each cluster)', required=False, default="default")

File: demos/iperf3/kind/simple_test.py
#!/usr/bin/env python3
# Copyright 2023 The ClusterLink Authors.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

################################################################
# Name: Simple iperf3  test
# Desc: create 2 kind clusters :
# 1) GW and iperf3 client
# 2) GW and iperf3 server    
###############################################################
import os
import sys

proj_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname( os.path.abspath(__file__)))))
sys.path.insert(0,f'{proj_dir}')

from demos.utils.common import runcmd, createFabric, printHeader, startGwctl
from demos.utils.kind import startKindCluster,useKindCluster, getKindIp,loadService

from demos.iperf3.kind.iperf3_client_start import directTestIperf3,testIperf3Client


############################### MAIN ##########################
if __name__ == "__main__":
    printHeader("\n\nStart Kind Test\n\n")
    printHeader("Start pre-setting")

    # GW parameters 
    gwPort           = "30443"
    gw1Name          = "peer1"
    gw2Name          = "peer2"
    srcSvc           = "iperf3-client"
    destSvc          = "iperf3-server"
    destPort         = 5000
    iperf3DirectPort = "30001"
    
    # Folders
    folCl=f"{proj_dir}/demos/iperf3/testdata/manifests/iperf3-client"
    folSv=f"{proj_dir}/demos/iperf3/testdata/manifests/iperf3-server"
    testOutputFolder = f"{proj_dir}/bin/tests/iperf3"

    # Policy
    allowAllPolicy=f"{proj_dir}/pkg/policyengine/policytypes/examples/allowAll.json"
    
    print(f'Working directory {proj_dir}')
    os.chdir(proj_dir)
    
    ### build docker environment 
    printHeader(f"Build docker image")
    os.system("make docker-build")
    os.system("sudo make install")
    
    ### Build Kind clusters environment 
    createFabric(testOutputFolder)
    startKindCluster(gw1Name, testOutputFolder)        
    startKindCluster(gw2Name, testOutputFolder)        
      
    ###get Gateways parameters
    gw1Ip = getKindIp(gw1Name)
    gw2Ip = getKindIp(gw2Name)
   
    # Start gwctl
    startGwctl(gw1Name, gw1Ip, gwPort, testOutputFolder)
    startGwctl(gw2Name, gw2Ip, gwPort, testOutputFolder)
    
    # Create peers
    printHeader("Create peers")
    runcmd(f'gwctl create peer --myid {gw1Name} --name {gw2Name} --host {gw2Ip} --port {gwPort}')
    runcmd(f'gwctl create peer --myid {gw2Name} --name {gw1Name} --host {gw1Ip} --port {gwPort}')
    
    # Set service iperf3-client in gw1
    loadService(srcSvc,gw1Name, "mlabbe/iperf3",f"{folCl}/iperf3-client.yaml" )
    runcmd(f'gwctl create export --myid {gw1Name} --name {srcSvc} --host {srcSvc} --port {destPort}')

    # Set service iperf3-server in gw2
    loadService(destSvc,gw2Name, "mlabbe/iperf3",f"{folSv}/iperf3.yaml" )
    runcmd(f"kubectl create service nodeport {destSvc} --tcp={destPort}:{destPort} --node-port={iperf3DirectPort}")
    runcmd(f'gwctl create export --myid {gw2Name} --name {destSvc} --host {destSvc} --port {destPort}')

    #Import destination service
    printHeader(f"\n\nStart Importing {destSvc} service to {gw1Name}")
    runcmd(f'gwctl --myid {gw1Name} create import --name {destSvc} --host {destSvc} --port {destPort}')
    printHeader(f"\n\nStart binding {destSvc} service to {gw1Name}")
    runcmd(f'gwctl --myid {gw1Name} create binding --import {destSvc} --peer {gw2Name}')

    #Add policy
    printHeader("Applying policies")
    runcmd(f'gwctl --myid {gw1Name} create policy --type access --policyFile {allowAllPolicy}')
    runcmd(f'gwctl --myid {gw2Name} create policy --type access --policyFile {allowAllPolicy}')
    
    #Testing
    printHeader("\n\nStart Iperf3 testing")
    directTestIperf3(gw1Name, srcSvc, gw2Ip, iperf3DirectPort)
    testIperf3Client(gw1Name, srcSvc, destSvc, destPort)


